typing/typecore.ml

```diff
@@ -4263,9 +4264,9 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
     let ls, tvar = list_labels env ty in
     not tvar && List.for_all ((=) Nolabel) ls
   in
-  match expand_head env ty_expected' with
-    {desc = Tarrow(Nolabel,ty_arg,ty_res,_); level = lv}
-    when is_inferred sarg ->
+  let lv = (repr ty_expected').level in
+  match expand_head env (correct_levels ty_expected') with
```

Comment:
I think this change raises a number of issues:
1. There are many other uses of `expand_head`, quite a few of which seem to cause the same issue.
2. Can we provide a better code pattern for handling these situations? Using `correct_levels` is not obvious. Maybe a function for specifically getting the head without risking changing the scope.
3. Can we avoid copying the whole type of all function arguments just to handle an extremely rare corner case? Can we delay copying until we detect that the expansion would change the scope? Can we avoid copying when there are no local constraints in the environment etc.
4. Can we push the functions that inspect types like this into Ctype? It seems better to avoid heavy type manipulation outside of `ctype.ml`. Anything which explicitly inspects the description of a type is "risky" and it would be nice to contain such code behind an abstraction boundary.

Reply:
* It have made the use of `correct_levels` conditional on `Env.local_constraints`, as it is only meaningful in that case.
* I have looked in `Typecore`, and I can indeed see a number of other suspicious calls to `Env.expand_head` (i.e., for inspection but not strictly required by unification). However, I would first like to find a counter-example before protecting those calls, as it is not even certain that the protection would be efficient.
* As for the need of an abstraction barrier, I agree, but it is currently pretty porous, and I would prefer not to mix this with this PR. I.e., it's probably better to keep semantical changes and refactoring in separate PRs.